kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/MatchersTest.kt

```diff
@@ -15,6 +22,36 @@ import kotlin.collections.HashMap
 
 class MatchersTest : FreeSpec({
 
+  "withClue()" - {
+    fun withClueEcho(other: String) = object : Matcher<String> {
+      override fun test(value: String)= Result(false,
+              "Should have the details of '$value' and $other",
+              "Should have the details of '$value' and $other")
+    }
+
+    "should prepend clue to message" {
+      val ex = shouldThrow<AssertionError> {
+        withClue("a clue:") { "1" shouldBe withClueEcho("here are the details!") }
+      }
+      ex.message shouldBe "a clue: Should have the details of '1' and here are the details!"
+    }
+
+    "should play nice with multiple/softAssert" {
+      val ex = shouldThrow<AssertionError> {
+        withClue("multi clue:") { assertSoftly {
+          "1" shouldBe withClueEcho("the details!")
+          "2" shouldBe withClueEcho("the details!")
+        } }
+      }
+      ex.message.apply {
+        shouldContain("multi clue:")
```

Comment:
Maybe it's better if we explicitliy test the message, similarly to the first test case. Testing for an exact value using `shouldBe`.

This test case doesn't cover a possible scenario of adding more than one clue, for example (I'm not even sure if this is the intended behavior, just as an example):
```
multi clue: Should have the details of '1' and the details!
multi clue: Should have the details of '2' and the details!
```

Reply:
Gah! You are correct. This doesn't work:

```
withClue("multi clue:") { assertSoftly {
          withClue("inner clue1") {"1" shouldBe withClueEcho("the details!") }
          withClue("inner clue2") {"2" shouldBe withClueEcho("the details!")}
        } }
```

The inner clues are ignored because the exceptions aren't thrown (and thus not caught). I'm going to investigate how the soft assertions work to see if I can integrate with it.

Also I used the `shouldContain` because the assertSoftly includes call site info:

```
java.lang.AssertionError: multi clue: 
The following 2 assertions failed:
1) Should have the details of '1' and the details!
	at com.sksamuel.kotlintest.matchers.MatchersTest$1$1$3$ex$1$1$1$1.invoke(MatchersTest.kt:42)
2) Should have the details of '2' and the details!
	at com.sksamuel.kotlintest.matchers.MatchersTest$1$1$3$ex$1$1$1$2.invoke(MatchersTest.kt:43)
 should include substring inner clue1: Should have the details of '1' and the details!
```

Reply:
Interesting.
It's OK to use `shouldContain`, but verify a specific amount is better IMO.

As for the test case itself using `assertSoftly`, I'm not sure what behavior is intended here. And I believe that this discussion should be in another ticket, as this one is not for this specific bug-report.

What do you think?